docs/design/gh-ost-integration.md

```diff
@@ -0,0 +1,474 @@
+
+```Go
+const TaskCheckDatabaseSchemaUpdateGhost = "bb.task-check.database.schema.update.ghost"
+type TaskCheckDatabaseSchemaUpdateGhostPayload struct {
+    Statement string `json:"statement,omitempty"`
+    InstanceID int `json:"instanceId,omitempty"`
+    DatabaseName string `json:"databaseName,omitempty"`
+    TableName string `json:"tableName,omitempty"`
+}
+```
+
+### In-migration progress
+
+gh-ost can report its migration progress. Migrator can write migration status to io.Writer. [Related codes here](https://github.com/github/gh-ost/blob/master/go/logic/migrator.go#L906).
+
+```Go
+// printStatus prints the progress status, and optionally additionally detailed
+// dump of configuration.
+// `rule` indicates the type of output expected.
+// By default the status is written to standard output, but other writers can
+// be used as well.
+func (this *Migrator) printStatus(rule PrintStatusRule, writers ...io.Writer) {
+}
+```
+
+It's important to show user the progress, because typically it will take hours or even days to complete when migrating on large tables. We periodically update the progress in `TaskRunResultPayload.Detail`.
+
+### Approve, run, approve, cut-over
+
+We will have two tasks for the actual migration. The first task runs gh-ost, syncs the ghost table with the original table. Once synced, the first task will be `DONE` and the second task will be `PENDING_APPROVAL`. The second task does the cut-over, which locks the tables and atomically switch table names. The second task is blocked by the first task.
+
+These two tasks need to access the same `Migrator`. Unfortunately, the way we schedule tasks now is not capable of it. Fortunately, we can leverage the unix socket file that gh-ost migrator listens on. When the ghost table catches up with the original table , the "run" task will switch to `DONE`, and the "cut over" task will be `PENDING_APPROVAL`. The "cut over" task sends `"cut-over"` to the socket file to execute cut over.
+
+The socket file is in `/tmp` directory, so the operating system will clean it up and we won't bother removing it. The socket file name will be `./tmp/gh-ost.{{DATABASE_NAME}}.{{TABLE_NAME}}.{{TIMESTAMP}}.sock`.
+
+We will have task checks for both tasks. For the "run" task, we have `TaskCheckDatabaseSchemaUpdateGhost` mentioned above. For the "cut-over" task, we will have a task check to check whether it is good to cut-over now (in sync, network lags good etc.)
+
+```Go
+const TaskCheckDatabaseSchemaUpdateGhostCutover = "bb.task-check.database.schema.update.ghost.cutover"
+type TaskCheckDatabaseSchemaUpdateGhostCutoverPayload struct {
+}
+```
+
+### Crash recovery
+
+If it crashes halfway when migrating on `tablename`, we will end up with the original `tablename`, and `_tablename_gho`, `_tablename_ghc`. We just drop `_tablename_gho` and `_tablename_ghc`, rerun gh-ost.
```

Comment:
not sure if we can suffix it with the timestamp.

though this might be a feature, since we can serve this as a gate during the precondition check. We disallow the migration if there exists _tablename_gho _tablename_ghc

Reply:
Let's create a separate section to list the pre-migration check items